File: retos/Reto-2/MiguelDecode/palindrome.js
const isWord = (string) => string.split(" ").length == 1;

const formatInput = (string) => {
  const regexSymbols = /[.!¡?¿''"";:~]/gi;

  const formatInput = string
    .trim()
    .toLowerCase()
    .replaceAll(" ", "")
    .replaceAll(regexSymbols, "")
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "");

  return formatInput;
};

const palindromo = (input) => {
  if (!input) throw new Error("No has introducido ninguna palabra o frase.");

  if (typeof input !== "string")
    throw new Error("El tipo de dato introducido no es un string.");

  if (isWord(input)) {
    // It's a word
    const word = formatInput(input);

    let reverseWord = word.split("").reverse().join("");

    return word === reverseWord
      ? "La palabra es un palíndromo"
      : "La palabra no es un palíndromo";
  } else {
    // It's a sentence
    const sentence = formatInput(input);

    let reverseSentence = sentence.split("").reverse().join("");

    return sentence === reverseSentence
      ? "La frase es un palíndromo"
      : "La frase no es un palíndromo";
  }
};

// console.log(palindromo());
// console.log(palindromo(33))
// console.log(palindromo([1, 2, 3]));

// console.log(palindromo("¿Acaso hubo búhos acá?"));
// console.log(palindromo("Ana"));
// console.log(palindromo("Frase random"));
// console.log(palindromo("Hola"));

File: retos/Reto-2/puriihuaman/Palindrome.java
import java.text.Normalizer;

public class Palindrome {
	private final static String REGEX = "[^a-z,^A-Z]";

	public static void main(String[] args) {

		System.out.println(isPalindromeWord("¿Acaso hubo búhos acá?"));
		System.out.println(isPalindromeWord("Ana"));
		System.out.println(isPalindromeWord("Otto"));
	}

	public static String isPalindromeWord(String word) {
		if (isAPhrase(word)) {
			return (isPalindrome(word)) ? "La frase es un palíndromo" : "La frase NO es un palíndromo";
		} else {
			return (isPalindrome(word)) ?
				"La palabra es un palíndromo" :
				"La palabra NO es un palíndromo";
		}
	}

	private static boolean isPalindrome(String text) {
		String currentText = normalizeString(text);
		String transformedText = normalizeString(new StringBuilder(currentText).reverse().toString());

		if (currentText.length() < 2) return false;
		else {
			return (currentText.equals(transformedText)) ? true : false;
		}
	}

	private static boolean isAPhrase(String text) {
		for (int i = 0; i < text.length(); i++) {
			if (text.charAt(i) == ' ') return true;
		}
		return false;
	}

	private static String normalizeString(String value) {
		value = Normalizer.normalize(value, Normalizer.Form.NFD);
		return value.replaceAll(REGEX, "").toLowerCase();
	}

	private static void palindromeInfo(String text, String type, boolean isTrue) {
		String informativeMessage = "";
		if (type.equalsIgnoreCase("word")) {
			if (isTrue) informativeMessage = "La palabra es un palíndromo";
			else informativeMessage = "La palabra no es un palíndromo";
		} else {
			if (isTrue) informativeMessage = "La frase es un palíndromo";
			else informativeMessage = "La frase no es un palíndromo";
		}
		System.out.println(informativeMessage);
	}
}
